src/pages/TokenDetails/Widget.tsx

```diff
@@ -0,0 +1,183 @@
+      } else {
+        setUserSlippage(new Percent(Math.floor(Number(slippage.max) * 100), 10_000))
+      }
+    },
+    [setUserSlippage]
+  )
+  const onSettingsReset = useCallback(() => {
+    setTtl(undefined)
+    setUserTtl(DEFAULT_DEADLINE_FROM_NOW)
+    setSlippage(undefined)
+    setUserSlippage('auto')
+  }, [setUserSlippage, setUserTtl])
+  const settings: SwapSettingsController = useMemo(() => {
+    const auto = userSlippage === 'auto'
+    return { slippage: { auto, max: slippage }, transactionTtl: ttl }
+  }, [slippage, ttl, userSlippage])
+  const settingsHandlers: SwapEventHandlers = useMemo(
+    () => ({ onSettingsReset, onSlippageChange, onTransactionDeadlineChange }),
+    [onSettingsReset, onSlippageChange, onTransactionDeadlineChange]
+  )
+
+  return { settings: { settings, ...settingsHandlers } }
+}
+
+// Integrates the Widget's controlled value, using the app's CurrencySearchModal for token selection.
+function useSyncWidgetValue(defaultToken: Currency) {
+  const [isExactInput, setIsExactInput] = useState(false)
+  const [amount, setAmount] = useState<string>()
+  const onAmountChange = useCallback((field: Field, amount: string) => {
+    setIsExactInput(field === Field.INPUT)
+    setAmount(amount)
+  }, [])
+  const [tokens, setTokens] = useState<{ [Field.INPUT]?: Currency; [Field.OUTPUT]?: Currency }>({
+    [Field.OUTPUT]: defaultToken,
+  })
+  const onSwitchTokens = useCallback(() => {
+    setIsExactInput((isExactInput) => !isExactInput)
+    setTokens((tokens) => ({
+      [Field.INPUT]: tokens[Field.OUTPUT],
+      [Field.OUTPUT]: tokens[Field.INPUT],
+    }))
+  }, [])
+  const [selectingField, setSelectingField] = useState<Field>()
+  const otherField = useMemo(() => (selectingField === Field.INPUT ? Field.OUTPUT : Field.INPUT), [selectingField])
+  const [selectingToken, otherToken] = useMemo(() => {
+    if (selectingField === undefined) return [undefined, undefined]
+    return [tokens[selectingField], tokens[otherField]]
+  }, [otherField, selectingField, tokens])
+  const onTokenSelectorClick = useCallback((field: Field) => {
+    setSelectingField(field)
+    return false
+  }, [])
+  const onTokenSelect = useCallback(
+    (token: Currency) => {
+      if (selectingField === undefined) return
+      setIsExactInput(selectingField === Field.INPUT)
+      setTokens(() => {
+        return {
+          [otherField]: token === otherToken ? selectingToken : otherToken,
+          [selectingField]: token,
+        }
+      })
+    },
+    [otherField, otherToken, selectingField, selectingToken]
+  )
+  const value: SwapController = useMemo(
+    () => ({ type: isExactInput ? TradeType.EXACT_INPUT : TradeType.EXACT_OUTPUT, amount, ...tokens }),
+    [amount, isExactInput, tokens]
+  )
+  const valueHandlers: SwapEventHandlers = useMemo(
+    () => ({ onAmountChange, onSwitchTokens, onTokenSelectorClick }),
+    [onAmountChange, onSwitchTokens, onTokenSelectorClick]
+  )
+
+  const tokenSelector = (
```

Comment:
does this need to get memo'd?

Reply:
i honestly don't know. because it is a component, I think it may be memoized by the react framework already.
I'll do some testing to figure this out.

Reply:
I don't think this needs to be memo-ized, but I don't think it will hurt.
I've memo-ized it by wrapping CurrencySearchModal in React.memo.